velocity/src/main/java/com/github/games647/fastlogin/velocity/listener/ConnectListener.java

```diff
@@ -0,0 +1,130 @@
+import com.github.games647.fastlogin.velocity.task.ForceLoginTask;
+import com.velocitypowered.api.event.Continuation;
+import com.velocitypowered.api.event.Subscribe;
+import com.velocitypowered.api.event.connection.DisconnectEvent;
+import com.velocitypowered.api.event.connection.PreLoginEvent;
+import com.velocitypowered.api.event.player.GameProfileRequestEvent;
+import com.velocitypowered.api.event.player.ServerConnectedEvent;
+import com.velocitypowered.api.proxy.InboundConnection;
+import com.velocitypowered.api.proxy.Player;
+import com.velocitypowered.api.proxy.server.RegisteredServer;
+
+import java.util.ArrayList;
+import java.util.UUID;
+import java.util.concurrent.TimeUnit;
+
+public class ConnectListener {
+
+    private final FastLoginVelocity plugin;
+    private final RateLimiter rateLimiter;
+
+    public ConnectListener(FastLoginVelocity plugin, RateLimiter rateLimiter) {
+        this.plugin = plugin;
+        this.rateLimiter = rateLimiter;
+    }
+
+    @Subscribe()
+    public void onPreLogin(PreLoginEvent preLoginEvent, Continuation continuation) {
+        if (!preLoginEvent.getResult().isAllowed()) {
+            return;
+        }
+        InboundConnection connection = preLoginEvent.getConnection();
+        if (!rateLimiter.tryAcquire()) {
+            plugin.getLog().warn("Simple Anti-Bot join limit - Ignoring {}", connection);
+            return;
+        }
+
+        String username = preLoginEvent.getUsername();
+        plugin.getLog().info("Incoming login request for {} from {}", username, connection.getRemoteAddress());
+
+        Runnable asyncPremiumCheck = new AsyncPremiumCheck(plugin, connection, username, continuation, preLoginEvent);
+        plugin.getScheduler().runAsync(asyncPremiumCheck);
+    }
+
+    @Subscribe
+    public void onGameprofileRequest(GameProfileRequestEvent event) {
+        if (event.isOnlineMode()) {
+            LoginSession session = plugin.getSession().get(event.getConnection().getRemoteAddress());
+
+            UUID verifiedUUID = event.getGameProfile().getId();
+            String verifiedUsername = event.getUsername();
+            session.setUuid(verifiedUUID);
+            session.setVerifiedUsername(verifiedUsername);
+
+            StoredProfile playerProfile = session.getProfile();
+            playerProfile.setId(verifiedUUID);
+            if (!plugin.getCore().getConfig().get("premiumUuid", true)) {
+                UUID offlineUUID = UUIDAdapter.generateOfflineId(playerProfile.getName());
+                event.setGameProfile(event.getGameProfile().withId(offlineUUID));
+                plugin.getLog().info("Overridden UUID from {} to {} (based of {}) on {}",
+                        verifiedUUID, offlineUUID, verifiedUsername, event.getConnection());
+            }
+
+            if (!plugin.getCore().getConfig().get("forwardSkin", true)) {
+                //FIXME: Do I need to remove *all* properties or only the skin related ones?
```

Comment:
Good point, only texture ones if that is possible. This should make it future proof.